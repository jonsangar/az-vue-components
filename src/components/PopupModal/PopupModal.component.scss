.popupModal{
  position: fixed;
  top: 0;
  left: 0;
  z-index: 100;
  width: 100vw;
  height: 100vh;

  /* Adding safe area  */
  padding-bottom: env(safe-area-inset-bottom);
  padding-bottom: var(--notch-inset-bottom);
  padding-left: env(safe-area-inset-left);
  padding-left: var(--notch-inset-left);
  background-color: rgba(0, 0, 0, 0.8);

  a {
    cursor: pointer;
  }

  table{
    tr:last-child{
      td{
        border: none;
      }
    }
  }

  &__backButton{
    position: absolute;
    top: simple();
    left: simple();
    cursor: pointer;
  }

  &__closeButton{
    position: absolute;
    top: half();
    right: 0.5rem;
    padding: 0 0.5rem 0;
    cursor: pointer;
  }

  &__wrapper {
    position: absolute;
    top: 0 ;
    right: 0;
    bottom: 0;
    left: 0;
    width: calc(100vw - #{simple()});
    height: fit-content;
    padding: 0;
    margin: auto;
    overflow: hidden;
    background-color: color(gray8);
    border-radius: $border-radius ;

  }

  &__container{
    max-height: 85vh;
    overflow-y: auto;
    scrollbar-width: none;

    &::-webkit-scrollbar{
      display: none;
    }
  }

  &__title {
    // @extend %base-bold-black;

    padding: simple() 0 simple();
    font-weight: bold;
    text-align: center;
  }

  &__body{
    .popupModal__img {
      display: block;
      margin: double() auto;
    }

    .popupModal__subtitle {
      margin-bottom: half();
      font-size: font-size(h4);
      font-weight: font-weight(bold);
      text-align: center;
      padding: 0 half();

      &--danger{
        color: color(danger);
      }

      &--brand{
        color: color(brand);
      }

      &--warning{
        color: color(warning);
      }
    }

    .popupModal__message {
      max-width: 320px;
      padding: 0 half();
      margin: 0 auto;
      margin-bottom: double();
      color: color(gray3);
      font-size: 1.1em;

      &--loading{
        padding-bottom: double();
        text-align: center;
      }

    }

    .popupModal__message--otherMessage {
      margin-bottom: 0 !important;
    }

    .popupModal__message--hasButtons{
      margin-bottom: quadruple();
    }

    .popupModal__list {
      width: 100%;
      padding: 0 half();

      &__container {
        list-style-type: none;
        padding: 0;
        border-radius: 10px;
        overflow: hidden;
      }

      &__item{
        position: relative;
        width: 100%;
        margin-left: 0;
        padding: half() simple();
        background-color: color(gray7);
        font-size: 1.1em;
        text-align: left;
        color: color(gray1);

        &:not(:last-child) {
          margin-bottom: 2px;
        }

        &__text {
          padding-right: calc(#{half()} + #{quarter()});
        }
        &__icon {
          position: absolute;
          top: 0;
          right: 0;
          padding: half();
          color: color(gray4);

        }

      }
    }

    .az-cell, .az-input, input{
      background-color: transparent;
    }

    .az-input, input{
      border-bottom: 1px solid color(gray6);
    }
  }


  &__buttons {
    position: absolute;
    bottom: 0;
    display: flex;
    width: 100%;
    padding-top: simple();
  }

  &__button {
    @extend %base-normal-black;

    width: 100%;
    height: double();
    overflow: hidden;
    line-height: double();
    color: color(gray8) !important;
    text-align: center;
    cursor: pointer;
    border-top: 1px solid $border-color;

    &:not(:last-child) {
      border-right: 1px solid $border-color;
    }
  }

  &__button--brand{
    background-image: radial-gradient(circle at 0% 0%, color(brand-light), color(brand));
  }

  &__button--cancel{
    background-image: radial-gradient(circle at 0% 0%, color(gray6), color(gray4));
  }

  &__footer{
    padding: half();
    font-size: font-size(small1);
    line-height: 1.5em;
  }

  &__backButton{
    width: 20%;
    padding: 0 0 half();
  }

  .linear-overlay-mask{
    position: absolute;
    bottom: 0;
    z-index: 2000;
    width: 100%;
    height: size(3);
    background: linear-gradient(0deg, rgba(255,255,255,1) 0%,rgba(255,255,255,1) 30%, rgba(255,255,255,0) 100%);
    border-radius: 0 0 $border-radius $border-radius;

    &--buttons {
      bottom: double();
    }
  }

  h1, h2, h3, h4, h5, h6{
    &.colour-title{
      padding-bottom: half();
      margin-bottom: half();
      font-weight: bold;

      &.color-1{
        color: #00AAA1;

        &.underline{
          border-bottom: 2px solid #00AAA1;
        }
      }

      &.color-2{
        color: #FFB837;

        &.underline{
            border-bottom: 2px solid #FFB837;
        }
      }

      &.color-3{
        color: #F3975B;

        &.underline{
            border-bottom: 2px solid #F3975B;
        }
      }

      &.color-4{
        color: #D04F53;

        &.underline{
          border-bottom: 2px solid #D04F53;
        }
      }

      &.color-5{
        color: #965379;

        &.underline{
          border-bottom: 2px solid #965379;
        }
      }
    }
  }

  .fade-enter-active, .fade-leave-active {
    transition: opacity .5s
  }
  .fade-enter, .fade-leave-to /* .fade-leave-active below version 2.1.8 */ {
    opacity: 0
  }
}

@include media(landscape) {
  .popupModal{
    &__wrapper {
      max-width: 370px;
    }
  }
}
