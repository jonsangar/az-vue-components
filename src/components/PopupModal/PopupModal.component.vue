<template lang="pug">
transition(name="fade-in-linear-transition")
  .popupModal(
    @click="closeOnClick && close($event)")
    transition(
      name="zoom-in-center-transition" appear)
      .popupModal__wrapper
        .popupModal__backButton(
          v-if="backButton"
          @click="cancel")
          az-icon(icon="left" scale="1"
            @click="cancel")
        .popupModal__closeButton(
          v-if="closeButton"
          @click="close")
          az-icon(icon="close" scale="1"
            @click="close")
        .popupModal__title.font-weight-bold {{title}}
        .popupModal__container(
          :id="`popup-${_uid}`")
          .popupModal__body
            img.popupModal__img(
              v-if="src"
              :src="require(`@/assets/${src}`)")
            .popupModal__subtitle(v-if="subtitle"
              :class="{'popupModal__subtitle--danger': danger}")
              | {{subtitle}}
            .popupModal__message(v-if="message"
              :style="textAlign ? `text-align: ${textAlign}` : ''"
              :class="{'popupModal__message--otherMessage': otherMessage, 'popupModal__message--hasButtons': buttons && !otherMessage}"
              @click="onClickMessage"
              v-html="message")
            .popupModal__message(v-if="otherMessage"
              :style="textAlign ? `text-align: ${textAlign}` : ''"
              :class="{'popupModal__message--hasButtons': buttons}"
              @click="onClickOtherMessage"
              v-html="otherMessage")
            .popupModal__list(
              v-else-if="listInfo && listInfo.length > 0"
              :style="textAlign ? `text-align: ${textAlign}` : ''"
              :class="{'popupModal__message--hasButtons': buttons}")
              ul.popupModal__list__container
                li.popupModal__list__item(
                  v-if="item.title"
                  v-for="item in listInfo")
                  .popupModal__list__item__text
                    | {{ item.title }}
                  .popupModal__list__item__icon
                    az-icon(
                      v-if="item.info"
                      scale=".55"
                      name="icon-info"
                      @click="showItemInfo(item)")
            .popupModal__message(
              v-else
              :style="textAlign ? `text-align: ${textAlign}` : ''"
              :class="{'popupModal__message--hasButtons': buttons}")
              slot
            .popupModal__message--loading(
              v-if="loading")
              az-loading(type="circle")
              .toast__title.toast__title--onlyTitle {{ loadingText }}

        transition(
          name="fade")
          .linear-overlay-mask.fade(
            :class="{'linear-overlay-mask--buttons': buttons}"
            v-show="showMask")

        .popupModal__buttons.margin-top(v-if="buttons")
          .popupModal__button.popupModal__button--cancel(
            v-if="cancelBtn"
            @click="cancel") {{ getBtnText(cancelBtnText) }}
          .popupModal__button.popupModal__button--brand(
            v-if="acceptBtn"
            @click="accept") {{ getBtnText(acceptBtnText) }}
        .popupModal__footer(v-else-if="$slots.buttons")
          slot(name="buttons")


</template>

<script>
import OverlayMask from 'Core/mixins/OverlayMask.mixin';
import openLink from 'Core/utils/openLink.utils';

export default {
  mixins: [
    OverlayMask
  ],

  props: {
    src: {
      type: String,
      default: ''
    },

    loading: {
      type: Boolean,
      default: false
    },

    loadingText: {
      type: String,
      default() {
        return this.$i18n ? this.$i18n.t('messages.loading') : '';
      }
    },

    title: {
      type: String,
      required: false
    },


    subtitle: {
      type: String,
      required: false
    },

    danger: {
      type: Boolean,
      required: false,
    },

    message: {
      type: String,
      required: false
    },

    otherMessage: {
      type: String,
      required: false
    },

    buttons: {
      type: Boolean,
      default: false
    },

    backButton: {
      type: Boolean,
      default: false
    },

    acceptBtn: {
      type: Boolean,
      default: true
    },

    cancelBtn: {
      type: Boolean,
      default: true
    },

    closeButton: {
      type: Boolean,
      default: false
    },

    cancelBtnText: {
      type: String,
      default() {
        return this.$i18n ? this.$i18n.t('buttons.cancel') : '';
      }
    },

    acceptBtnText: {
      type: String,
      default() {
        return this.$i18n ? this.$i18n.t('buttons.accept') : '';
      }
    },

    textAlign: {
      type: String,
      default: 'center'
    },

    closeOnClick: {
      type: Boolean,
      default: true
    },

    /**
     * Si está a true evita entrar en el "catch" al cancelar pulsando sobre la máscara en una instancia del Popup
     */
    preventOnClickMaskReject: {
      type: Boolean,
      default: false,
    },

    listInfo: {
      type: Array,
      default: () => {
        return []
      }
    },

    callback: Function,

  },

  methods: {

    getBtnText(btnText){
      if(typeof btnText === 'function'){
        return btnText();
      }

      return btnText;
    },

    accept(){

      if(this.callback) {
        this.callback('confirm');
        this.clear();
      }


      this.$emit('accept');

    },

    close(e) {
      if(e.target !== e.currentTarget) return;

      if(this.callback) {
        if(!this.preventOnClickMaskReject) {
          this.callback('cancel');
        }
        this.clear();
      }

      this.$emit('close');
    },

    cancel(e) {
      if(e.target !== e.currentTarget) return;

      if(this.callback) {
        this.callback('cancel');
        this.clear();
      }

      this.$emit('cancel');
    },

    showItemInfo(item) {
      this.$emit('showInfo',item);
    },

    onClickMessage(e) {
      // console.log('onClickMessage: ', e)
      if (e?.target?.href){
        e.preventDefault()
        openLink(e.target.href)
      }
    },

    onClickOtherMessage(e) {
      // console.log('onClickOtherMessage: ', e)
      if (e?.target?.href){
        e.preventDefault()
        openLink(e.target.href)
      }
    },

    calculateOverlayMaskPosition() {
      this.$nextTick(() => {
        if(this.$slots.buttons){
          // Si usamos footer personalizado a través de slot, la altura puede variar. Se calcula para ubicarlo en la posición
          const mask = document.querySelector('.linear-overlay-mask');
          const footer = document.querySelector('.popupModal__footer');
          mask.style.bottom = `${footer.clientHeight}px`;
          mask.style.borderBottomLeftRadius = 0;
          mask.style.borderBottomRightRadius = 0;
        }
      })
    }
  },

  created() {
    this.initOverlaySettings(`popup-${this._uid}`);
    this.calculateOverlayMaskPosition();
  },

  mounted() {
    document.body.classList.add('popup-overflow');
    document.body.appendChild(this.$el);
  },

  beforeDestroy() {
    document.body.classList.remove('popup-overflow');

    const parent = this.$el.parentNode;
    if(parent) parent.removeChild(this.$el);
  }

}
</script>

<style lang="scss" src="./PopupModal.component.scss" scoped></style>
