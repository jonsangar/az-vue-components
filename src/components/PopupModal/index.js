import PopupModal from './PopupModal.component.vue';

export default PopupModal;
