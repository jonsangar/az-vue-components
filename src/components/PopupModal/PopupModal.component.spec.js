import { mount, createLocalVue } from '@vue/test-utils';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import PopupModal from './PopupModal.component.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('PopupModal.vue', () => {
  const build = () => {
    const wrapper = mount(PopupModal, {
      localVue,
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente PopupModal con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Comprueba que el título se renderiza cuando se asigna la prop', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      title: 'Warnings',
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.popupModal__title').text()).toBe('Warnings')
  })

  it('Comprueba que el subtítulo se renderiza cuando se asigna la prop', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      subtitle: 'MySubtitle',
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.popupModal__subtitle').text()).toBe('MySubtitle')
  })

  it('Comprueba que el mensaje se renderiza cuando se asigna la prop', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'MyMessage',
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.popupModal__message').text()).toBe('MyMessage')
  })

  it('Comprueba que dos mensajes se renderizan cuando se asignan las props', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'Message',
      otherMessage: 'OtherMessage',
    })
    await wrapper.vm.$nextTick()
    const messages = wrapper.findAll('.popupModal__message')
    expect(messages.length).toBe(2)
    expect(messages.at(0).text()).toBe('Message')
    expect(messages.at(1).text()).toBe('OtherMessage')
  })

  it('Comprueba que no se renderiza los botones por defecto si no se asignan las props', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'Message',
      otherMessage: 'OtherMessage',
    })
    await wrapper.vm.$nextTick()
    const buttons = wrapper.findAll('.popupModal__buttons')
    expect(buttons.length).toBe(0)
  })

  it('Comprueba que los dos botones se renderizan cuando se asignan las props', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'Message',
      otherMessage: 'OtherMessage',
      buttons: true,
    })
    await wrapper.vm.$nextTick()
    const buttons = wrapper.findAll('.popupModal__button')
    expect(buttons.length).toBe(2)
  })

  it('Comprueba que los dos textos de los botones se renderizan cuando se asignan las props', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'Message',
      otherMessage: 'OtherMessage',
      buttons: true,
      acceptBtn: true,
      cancelBtn: true,
      acceptBtnText: 'Aceptar',
      cancelBtnText: 'Cancelar',
    })
    await wrapper.vm.$nextTick()
    const buttons = wrapper.findAll('.popupModal__button')
    expect(buttons.length).toBe(2)
    expect(buttons.at(0).text()).toBe('Cancelar')
    expect(buttons.at(1).text()).toBe('Aceptar')
  })

  it('Comprueba que solo se renderiza el texto del botón aceptar cuando se asignan las props', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'Message',
      otherMessage: 'OtherMessage',
      buttons: true,
      acceptBtn: true,
      cancelBtn: false,
      acceptBtnText: 'Aceptar',
    })
    await wrapper.vm.$nextTick()
    const buttons = wrapper.findAll('.popupModal__button')
    expect(buttons.length).toBe(1)
    expect(buttons.at(0).text()).toBe('Aceptar')
  })

  it('Comprueba que solo se renderiza el texto del botón cancelar cuando se asignan las props', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'Message',
      otherMessage: 'OtherMessage',
      buttons: true,
      acceptBtn: false,
      cancelBtn: true,
      cancelBtnText: 'Cancelar',
    })
    await wrapper.vm.$nextTick()
    const buttons = wrapper.findAll('.popupModal__button--cancel')
    await buttons.trigger('click')
    // Compruebo que existe el botón con el texto
    expect(buttons.length).toBe(1)
    expect(buttons.at(0).text()).toBe('Cancelar')
  })

  it('Comprueba que emite cuando hay info', async () => {
    const { wrapper } = build()
    const item = [{title: 'MyTitle', info: 'warning'}]
    wrapper.setProps({
      title: 'Warnings',
      subtitle: 'MySubtitle',
      listInfo: item,
    })
    await wrapper.vm.$nextTick()
    wrapper.vm.showItemInfo(item)
    expect(wrapper.emitted().showInfo).toBeTruthy()
    expect(wrapper.emitted().showInfo.length).toBe(1)
    expect(wrapper.emitted().showInfo[0]).toEqual([item])
  })

  it('Comprueba que emite cuando se cancela', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'Message',
      otherMessage: 'OtherMessage',
      buttons: true,
      acceptBtn: false,
      cancelBtn: true,
      cancelBtnText: 'Cancelar',
    })
    await wrapper.vm.$nextTick()
    const buttons = wrapper.findAll('.popupModal__button--cancel')
    await buttons.trigger('click')
    // Compruebo que emite al ser clicado
    expect(wrapper.emitted().cancel).toBeTruthy()
    expect(wrapper.emitted().cancel.length).toBe(1)
    expect(wrapper.emitted().cancel[0]).toEqual([])
  })

  it('Comprueba que emite cuando se cierra', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'Message',
      otherMessage: 'OtherMessage',
      buttons: true,
    })
    await wrapper.vm.$nextTick()
    const buttons = wrapper.findAll('.popupModal')
    await buttons.trigger('click')
    // Compruebo que emite al ser clicado
    expect(wrapper.emitted().close).toBeTruthy()
    expect(wrapper.emitted().close.length).toBe(1)
    expect(wrapper.emitted().close[0]).toEqual([])
  })

  it('Comprueba que emite cuando se acepta', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      message: 'Message',
      otherMessage: 'OtherMessage',
      buttons: true,
      acceptBtn: true,
    })
    await wrapper.vm.$nextTick()
    const buttons = wrapper.findAll('.popupModal__button')
    await buttons.trigger('click')
    // Compruebo que emite al ser clicado
    expect(wrapper.emitted().accept).toBeTruthy()
    expect(wrapper.emitted().accept.length).toBe(1)
    expect(wrapper.emitted().accept[0]).toEqual([])
  })

  it('Comprueba que emite cuando el botón de texto es una función', async () => {
    const { wrapper } = build()
    const returned = await wrapper.vm.getBtnText('function')
    expect(returned).toBe('function')
  })
});
