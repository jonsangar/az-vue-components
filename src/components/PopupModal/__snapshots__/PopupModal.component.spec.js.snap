// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupModal.vue Comprueba renderizado del componente PopupModal con snapshot 1`] = `
<transition-stub name="fade-in-linear-transition">
  <div class="popupModal">
    <transition-stub name="zoom-in-center-transition" appear="true">
      <div class="popupModal__wrapper">
        <!---->
        <!---->
        <div class="popupModal__title font-weight-bold"></div>
        <div id="popup-1" class="popupModal__container">
          <div class="popupModal__body">
            <!---->
            <!---->
            <!---->
            <div class="popupModal__message" style="text-align: center;"></div>
            <!---->
          </div>
        </div>
        <transition-stub name="fade">
          <div class="linear-overlay-mask fade" style="display: none;"></div>
        </transition-stub>
        <!---->
      </div>
    </transition-stub>
  </div>
</transition-stub>
`;
